dashboard/client/src/pages/job/TaskTimeline.tsx

```diff
@@ -0,0 +1,231 @@
+  taskProgressTable: {
+    marginTop: theme.spacing(2),
+  },
+}));
+
+// The module guarantees there's only one fetch at
+// a time meaning there can be only one timer at a time.
+let timer: null | any = undefined;
+
+export const TaskTimeline = ({ jobId = null }: { jobId: string | null }) => {
+  const classes = useStyle();
+
+  const TimelineDownloadButton = ({
+    jobId = null,
+  }: {
+    jobId: string | null;
+  }) => {
+    const [loading, setLoading] = useState(false);
+
+    const onClick = () => {
+      // Download the chrome tracing file.
+      if (!loading) {
+        setLoading(true);
+        // blob response type is necessary to
+        DownloadTaskTimeline(jobId)
+          .then((blob) => {
+            const url = window.URL.createObjectURL(blob.data);
+            const a = document.createElement("a");
+            a.href = url;
+            a.download = `timeline-${dayjs().format()}-${jobId}.json`;
+            a.click();
+            a.remove();
+          })
+          .catch((error) => {
+            console.log(
+              `Failed to fetch data for the timeline. Error: ${error}`,
+            );
+          })
+          .finally(() => {
+            setLoading(false);
+          });
+      }
+    };
+
+    if (loading) {
+      return (
+        <div>
+          <Button variant="outlined" disabled={loading}>
+            Download Timeline <CircularProgress color="primary" size={16} />
+          </Button>
+        </div>
+      );
+    } else {
+      return (
+        <Button variant="outlined" disabled={loading} onClick={onClick}>
+          Download Timeline
+        </Button>
+      );
+    }
+  };
+
+  const TimelineButton = ({ jobId = null }: { jobId: string | null }) => {
+    const [loading, setLoading] = useState(false);
+    const [openDialog, setOpenDialog] = useState(false);
+
+    // Code from https://perfetto.dev/docs/visualization/deep-linking-to-perfetto-ui
+    const openTrace = (arrayBuffer: any) => {
+      const win = window.open(PERFETTO_URL);
+      const start = new Date().getTime();
+      if (win) {
+        window.addEventListener("message", (evt) =>
+          onMessage(evt, win, arrayBuffer, start),
+        );
+        timer = setInterval(() => win.postMessage("PING", PERFETTO_URL), 50);
+      }
+    };
+
+    const onMessage = (evt: any, win: any, arrayBuffer: any, start: any) => {
+      if (new Date().getTime() - start > PERFETTO_LOADING_TIMEOUT_MS) {
+        // If perfetto doesn't respond in 5 seconds, fail it.
+        setLoading(false);
+        return;
+      }
+
+      if (evt.data !== "PONG") {
+        return;
+      }
+
+      window.clearInterval(timer);
+      win.postMessage(arrayBuffer, PERFETTO_URL);
+      // We now finished uploading timeline.
+      setLoading(false);
+    };
+
+    const onDialogClose = () => {
+      setOpenDialog(false);
+    };
+
+    const onClick = () => {
+      setOpenDialog(true);
+    };
+
+    const onClickPerfettoLoading = () => {
```

Comment:
```suggestion
    const handleDialogClose = () => {
      setOpenDialog(false);
    };

    const handleClick = () => {
      setOpenDialog(true);
    };

    const handleClickPerfettoLoading = () => {
```

nit: naming convention